proposals/0008-c-interface-compiler-library.md

```diff
@@ -996,21 +998,535 @@ if (compileResult->HasOutput(DXC_OUT_OBJECT)) {
+or other state output from the compiler.
+
+```c++
+/**
+ * Error codes returned by library routines.
+ *
+ * Zero (\c HlslError_Success) is the only error code indicating success.
+ * Other error codes, including not yet assigned non-zero values, indicate
+ * errors.
+ * 
+ * Check CompilerLibraryResult for additional details from library 
+ * operations. This will include additional stdout/stderr outputs
+ * that occured during the call.
+ */
+enum CompilerLibraryErrorCode {
+    /**
+     * No error.
+     */
+    CompilerLibraryError_Success = 0,
+
+    /**
+     * A generic error code.
+     *
+     * Errors of this kind can get their own specific error codes in future
+     * library versions.
+     */
+    CompilerLibraryError_Failure = 1,
+
+    /**
+     * The operation is currently in progress.
+     */
+    CompilerLibraryError_Busy = 2,
+
+    /**
+     * The operation has been canceled.
+     */
+    CompilerLibraryError_Canceled = 3,
+
+    /**
+     * The function detected that the arguments violate the function
+     * contract.
+     */
+    CompilerLibraryError_InvalidArguments = 4,
+
+    /**
+     * The function detected that the specified cookie value
+     * is invalid.
+     */
+    CompilerLibraryError_InvalidCookie = 5,
+
+    /**
+     * The timeout value for waiting for an operation to be completed
+     * has been reached.
+     */
+    CompilerLibraryError_WaitTimeout = 6,
+};
+
+/**
+ * Extended error and result information collected during library
+ * operations.
+ */
+struct CompilerLibraryResult
+{
+    /**
+     * An error code indicating success or otherwise.
+     */
+    CompilerLibraryErrorCode error_code;
+
+    /**
+     * An opaque type that matches the cookie associated
+     * with the operation.
+     */
+    CompilerLibraryCookie operation_cookie;
+
+    /**
+     * Null terminated text buffer containing the captured
+     * stdout traffic output during compilation.
+     */
+    const char* stdout_text;
+
+    /**
+     * Size of the stdout_text buffer in bytes.
+     */
+    int stdout_size_bytes;
+
+    /**
+     * Null terminated text buffer containing the captured
+     * stderr traffic output during compilation.
+     */
+    const char* stderr_text;
+
+    /**
+     * Size of the stdout_text buffer in bytes.
+     */
+    int stderr_size_bytes;
+};
 
-### Alternative 2: Build/Ship Nuget package library
+/**
+* Retrieve additonal compilation result information.
+*
+* A return value of false indicates there are no additional results to 
+* retrieve.
+* 
+* \param instance the compiler instance
+*
+* \param cookie an opaque value returned from clang_compile_async that
+* identifies the operation whos results we want to retrieve.
+* 
+* \param result additional information for the compile operation.
+* clang_disposeCompilerLibraryResult must be called to free allocated
+* result memory.
+*/
+bool clang_getCompilerLibraryResult(
```

Comment:
There's operation_cookie in CompilerLibraryResult already, why need another cookie parameter in clang_getCompilerLibraryResult? For case result is nullptr?

Reply:
The cookie is to retrieve the specific result of the api call.  Having the cookie value mirrored in the result data was a convenience thing to be able to match results to their calls.  Some more thinking might need to be done here.